### DISCUSSION TITLE
Beatmap format specification

### DISCUSSION THREAD

**Discussion by fmang** [ppy/osu#12976] [Ideas]
**TL;DR: The beatmap format is more complex than you'd imagine, and there's room for improvement. But first, let's document it, because it's far from obvious.**

I've been recently working on a lightweight osu port for slow computers ([fmang/oshu](https://github.com/fmang/oshu)) and had to work on a [beatmap parser](https://github.com/fmang/oshu/blob/master/src/beatmap/beatmap.h). All I've found so far is this page on the wiki https://osu.ppy.sh/help/wiki/osu!_File_Formats/Osu_(file_format)

While at first you'd think it's almost self-explanatory, there are tons of unclear or completely weirds points. Here are a few:

1. The unit for slider duration is weird. It's like 100 means 1 beat then you multiply it by the duration per beat of the timing point and then the slider multiplier, while it'd be so much simpler if it were just the duration in milliseconds.

2. The circle size, and more generally the difficulty settings are in a completely obscure unit, where 5 is average and get harder as it increases. I first expected the circle size to be in pixels, and that a circle size of 9 were bigger than 5, but that's the opposite.

3. One of field in the timing points is the custom index of the sample it seems, while the custom index in the additions is not used anywhere anymore. That point is completely undocumented.

4. The milliseconds per beat is in milliseconds, or when negative it's actually a percentage of the previous non-negative timing point..? Then what is the inherited flag for? Beside, how is storing negative values any more efficient that using the computed value in milliseconds?

5. 0 sometimes mean “pick the default thing”, something it's an actual 0. For the sound type, 0 means 1 and there's no way to make the note mute. For the sample index, 0 means 1 too I believe. For the volume I believe 0 means mute? To identify the sample set (soft, normal, drum), 0 means “use the sample set of the timing point, which itself might be 0 but then you'd use the beatmap's one”. Why the fancy logic? Writing a straight value without relying on inheritance is much easier.

6. For Bézier paths, repeated points represent the end of a segment and the start of a new one. However, it's totally okay in theory to have a cubic curve where your control points are A, B, B, C. That path can't be written in a beatmap because it would be parsed as two straight lines AB and BC.

7. The background image is specified in the [Events] section in a weird CSV-like format, between quotes. What if the filename contains a double quote? Or a comma?

I've had to do a lot of searching to find some of these points, and if osu weren't open source, I don't know if I could have made a working parser.

Writing out a detailed specification would make edge cases stand out, and open the way for improvements. To ensure it's correct, I believe support from an official developer is required.

If I could get the help from one of you, I'd be willing to contribute to, or even lead that initiative.

**Comment by peppy** [comment]
The source code is already available here. If you want to update the wiki sepc, that would be very much appreciated. If you want to ask questions, there's discord for that.

We are planning on making a new beatmap format once lazer is in a stable state, but if you want to start on a proposal spec for that I won't stop you.

**Comment by fmang** [comment]
Alright! I'll check out that discord thing.

I have a few ideas for a proposal spec. Where would you like me to post it?

**Comment by Speykious** [comment]
Hello!

I dug straight down while searching for someone who had potentially voiced my concern and I found this thread from 2017.

For a while now, I've had this thought that the current osu! file format (including v128) is overengineered, and some of the ways it works defeats the point of having it under a readable form. So I was thinking of participating in the proposal of ideas for a new format that would be a major breaking change from the last one (as I understand that v128 is backwards-compatible with v14).

Is there already an issue, discussion or even pull request on that proposal?

(For the record, [I also wrote my own osu! beatmap file parser](https://github.com/Speykious/oSUS) in Rust.)

## Edit

I just got my answer on Discord from frenzibyte.
This is the only discussion tracking it, damn...
Well I'm glad to be early in the discussion about it then. xD

I may come back with a proposal for a new beatmap file format later if I have enough inspiration. It's very low priority right now though so I won't expect devs to spend too much time thinking about it.
